docs/architecture/design-proposals/dashboard/deployment-improvements.md

```diff
@@ -0,0 +1,197 @@
+# Improving User Experience for Deploying and Upgrading Applications
+
+The goal of this document is to specify the design proposal for improving the current approach for deploying and upgrading applications.
+
+## Definition of the problem
+
+When users deploy an application we present the default `values.yaml` that the chart contains. This is the only mechanism they have to modify the default behavior. In many cases, they want to modify this default setup with simple modifications like changing the default hostname or modify the default password. It's not trivial to find these settings in the values.yaml that it's usually very verbose. They should be able to modify some "common" parameters easily. In any case, for advanced users, they should be able to still modify the `values.yaml` to fine-tune the settings of the application.
+
+Once an application is deployed, eventually the user will need to upgrade it. Now, apart from the complexity of having to handle the `values.yaml` file, it's necessary to remember the changes done in the original `values.yaml` in order to port those to the new version. In Kubeapps, once a user selects a new version, the values are substituted with the defaults for the new version, making it very difficult to modify these values with the previous one. It should be possible to follow the same approach than in the deployment: have an easy-to-fill form, having the possibility to switch to the advanced form in which they can modify the values as their liking.
+
+## Desired User Experience (UX)
+
+### Deployment and Upgrade Form
+
+When users reach the deployment form, the desired behavior is to have two different views, one for the basic form and another for an advanced deployment. That can be represented as two different tabs. The basic deployment should contain all the basic parameters while the advanced form will show the information as we do today. For the first version, changes made in the basic deployment form won't affect the advanced form and viceversa. Only when the user clicks on "submit" the current form information (either basic or advanced) will be sent to the backend.
```

Comment:
Agree with @absoludity this is an important feature. As AWS does with Visual / JSON editors, changes should be reflected on both options. I can start filling the basic form and them switch to the advanced one to change something.

Without entering on implementation details, I believe we can define a mapping between the basic form and the `values.yaml` keys. I agree that replacing content may introduce unexpected errors. However, by defining this mapping, we can take the value from the `vaules.yaml` and print it on the basic form and viceversa.

Reply:
That sounds doable but not sure if at the end it will be. I am mostly worried about order, comments and multiple spaces that's why I didn't want to add a compromise here, but if you think it's a must-have I can add and see what we can do.

Reply:
> I haven't tried it, but from https://github.com/eemeli/yaml/ : "As a particularly distinguishing feature, yaml fully supports reading and writing comments and blank lines in YAML documents." . We could do a quick test to see if it parses Kubeapss' values.yaml and restringifies it without a diff?

Apparently, it works as advertised. The only issue I have find is that if a comment has blank lines within it, the library removes them:

```
> a
'foo: bar\n\n# THIS IS A COMMENT\n\n# Other comment\nbar:\n  foobar: barfoo\n\n'
> c = yaml.parseDocument(a)
Document {
...
> c.set("foo", "not-foo")
> c.toString()
'foo: not-foo\n\n# THIS IS A COMMENT\n# Other comment\nbar:\n  foobar: barfoo\n'
```

In any case, it's good enough so let's go with that.

Reply:
Excellent. I assume the parsed value easy to work with?